src/librustc_typeck/check/demand.rs

```diff
@@ -58,7 +60,34 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         if let Err(e) = self.try_coerce(expr, checked_ty, expected) {
             let origin = TypeOrigin::Misc(expr.span);
             let expr_ty = self.resolve_type_vars_with_obligations(checked_ty);
-            self.report_mismatched_types(origin, expected, expr_ty, e);
+            let mut diag = self.mismatched_types_diag(origin, expected, expr_ty, e.clone());
+
+            if let Node::NodeBlock(block) = self.tcx.map.get(self.tcx.map.get_parent_node(expr.id)) {
+                if let TypeError::Sorts(ref values) = e {
+                    if values.expected.is_nil() {
+                        // an implicit return to a method with return type `()`
+                        diag.span_label(expr.span,
+                                        &"possibly missing `;` here?");
+                        // get the current node's method definition
+                        if let Node::NodeExpr(item) = self.tcx.map
+                            .get(self.tcx.map.get_parent_node(block.id))
+                        {
+                            if let Node::NodeItem(func) = self.tcx.map
+                                .get(self.tcx.map.get_parent_node(item.id))
+                            {
```

Comment:
I don't like this, it's too arbitrary. If you want to do this properly you'd have to add something to `Expectation` to indicate that the type comes from a function return type.